Bug in variable scope

When re-defining a constant variable in an inner scope, we have an compilation error.
Example: given

``` golo
module ReferenceScopeTest

function main = |args| {
  let a = 42
  if true {
    let a = 21
    println(a)
  }
  println(a)
}
```

we have:

```
[error] Assigning `a` at {line=6, column=5} but it is a constant reference
```

while the expected behavior is

```
21
42
```

The compiler message is a bit confusing but the current behaviour seems to me legal.

```java
public class Test {

  public static void main(String[] args){
    int a = 42;
    if(true) {
      int a = 21;
      System.out.println(a);
    }
    System.out.println(a);
  }

}
```

for this, the java compiler says: 

```
Test.java:6: error: variable a is already defined in method main(String[])
```

Maybe it is possible to add the behaviour you describe but i'm not sure it's good thing for the readability.

When Yannick showed me this bug last Friday I thought it was indeed a scoping issue, but on second thought and with Sylvain’s example from Java I am now on the side that this is not a bug.

But at least the compiler error could be more precise and explain that shadowing a constant is not a good thing to do.

On 24 April 2016 at 09:48:54, sylvain Desgrais (notifications@github.com) wrote:
> The compiler message is a bit confusing but the current behaviour seems to me legal.
>  
> ```java
> public class Test {
>  
> public static void main(String[] args){
> int a = 42;
> if(true) {
> int a = 21;
> System.out.println(a);
> }
> System.out.println(a);
> }
>  
> }
> ```
>  
> for this, the java compiler says:
>  
> ```
> Test.java:6: error: variable a is already defined in method main(String[])
> ```
>  
> Maybe it is possible to add the behaviour you describe but i'm not sure it's good thing  
> for the readability.
>  
>  
>  
> ---
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/eclipse/golo-lang/issues/380#issuecomment-213909111

--- a/src/main/java/org/eclipse/golo/compiler/LocalReferenceAssignmentAndVerificationVisitor.java
+++ b/src/main/java/org/eclipse/golo/compiler/LocalReferenceAssignmentAndVerificationVisitor.java
@@ -140,16 +140,16 @@ class LocalReferenceAssignmentAndVerificationVisitor extends AbstractGoloIrVisit
   public void visitAssignmentStatement(AssignmentStatement assignmentStatement) {
     LocalReference reference = assignmentStatement.getLocalReference();
     Set<LocalReference> assignedReferences = assignmentStack.peek();
-    if (assigningConstant(reference, assignedReferences)) {
-      getExceptionBuilder().report(ASSIGN_CONSTANT, assignmentStatement.getASTNode(),
-          "Assigning `" + reference.getName() +
-              "` at " + assignmentStatement.getPositionInSourceCode() +
-              " but it is a constant reference"
-      );
-    } else if (redeclaringReferenceInBlock(assignmentStatement, reference, assignedReferences)) {
+    if (redeclaringReferenceInBlock(assignmentStatement, reference, assignedReferences)) {
       getExceptionBuilder().report(REFERENCE_ALREADY_DECLARED_IN_BLOCK, assignmentStatement.getASTNode(),
-          "Declaring a duplicate reference `" + reference.getName() +
-              "` at " + assignmentStatement.getPositionInSourceCode()
+          "Declaring a duplicate reference `" + reference.getName()
+          + "` at " + assignmentStatement.getPositionInSourceCode()
+      );
+    } else if (assigningConstant(reference, assignedReferences)) {
+      getExceptionBuilder().report(ASSIGN_CONSTANT, assignmentStatement.getASTNode(),
+          "Assigning `" + reference.getName()
+          + "` at " + assignmentStatement.getPositionInSourceCode()
+          + " but it is a constant reference"
       );
     }
     bindReference(reference);
@@ -178,8 +178,8 @@ class LocalReferenceAssignmentAndVerificationVisitor extends AbstractGoloIrVisit
 
   private boolean assigningConstant(LocalReference reference, Set<LocalReference> assignedReferences) {
     return reference.isConstant() && (
-        assignedReferences.contains(reference) ||
-        (reference.isModuleState() && !functionStack.peek().isModuleInit()));
+        assignedReferences.contains(reference)
+        || (reference.isModuleState() && !functionStack.peek().isModuleInit()));
   }
 
   private boolean referenceNameExists(LocalReference reference, Set<LocalReference> referencesInBlock) {

